### ISSUE TITLE
heap-use-after-free at theory_str.cpp:2345 (z3str3)

### ISSUE THREAD

**Issue by rainoftime** [Z3Prover/z3#4294]
[2345.txt](https://github.com/Z3Prover/z3/files/4613638/2345.txt)

z3 (commit a14c2a3) throws a heap-use-after-free
~~~~
=================================================================
==61446==ERROR: AddressSanitizer: heap-use-after-free on address 0x60e00000f2d0 at pc 0x00000141d901 bp 0x7fc296f40740 sp 0x7fc296f40730
READ of size 8 at 0x60e00000f2d0 thread T1
    #0 0x141d900 in smt::theory_str::simplify_parent(expr*, expr*) ../src/smt/theory_str.cpp:2345
    #1 0x14535b4 in smt::theory_str::handle_equality(expr*, expr*) ../src/smt/theory_str.cpp:6705
    #2 0x1457802 in smt::theory_str::new_eq_eh(int, int) ../src/smt/theory_str.cpp:6992
    #3 0xfc5a08 in smt::context::propagate_th_eqs() ../src/smt/smt_context.cpp:1617
    #4 0xfc65bf in smt::context::propagate() ../src/smt/smt_context.cpp:1694
    #5 0xfdcae0 in smt::context::bounded_search() ../src/smt/smt_context.cpp:3732
    #6 0xfdacdc in smt::context::search() ../src/smt/smt_context.cpp:3613
    #7 0xfd9269 in smt::context::check(unsigned int, expr* const*, bool) ../src/smt/smt_context.cpp:3496
    #8 0xd3f814 in smt::kernel::imp::check(unsigned int, expr* const*) ../src/smt/smt_kernel.cpp:116
    #9 0xd3e471 in smt::kernel::check(unsigned int, expr* const*) ../src/smt/smt_kernel.cpp:296
    #10 0x10f36b6 in check_sat_core2 ../src/smt/smt_solver.cpp:190
    #11 0x19e4fcb in solver_na2as::check_sat_core(unsigned int, expr* const*) ../src/solver/solver_na2as.cpp:67
    #12 0x19e9035 in solver::check_sat(unsigned int, expr* const*) ../src/solver/solver.cpp:330
    #13 0x5e0734 in solver::check_sat(ref_vector<expr, ast_manager> const&) ../src/solver/solver.h:149
    #14 0x19f84e5 in parallel_tactic::solver_state::simplify() ../src/solver/parallel_tactic.cpp:265
    #15 0x19fb07e in parallel_tactic::cube_and_conquer(parallel_tactic::solver_state&) ../src/solver/parallel_tactic.cpp:481
    #16 0x19fc433 in parallel_tactic::run_solver() ../src/solver/parallel_tactic.cpp:634
    #17 0x19fcb0f in parallel_tactic::solve(ref<model>&)::{lambda()#1}::operator()() const (/home/peisen/test/tofuzz/z3-debug/build/z3+0x19fcb0f)
    #18 0x1a02c6f in void std::_Bind_simple<parallel_tactic::solve(ref<model>&)::{lambda()#1} ()>::_M_invoke<>(std::_Index_tuple<>) /usr/include/c++/5/functional:1531
    #19 0x1a02b83 in std::_Bind_simple<parallel_tactic::solve(ref<model>&)::{lambda()#1} ()>::operator()() /usr/include/c++/5/functional:1520
    #20 0x1a02b13 in std::thread::_Impl<std::_Bind_simple<parallel_tactic::solve(ref<model>&)::{lambda()#1} ()> >::_M_run() /usr/include/c++/5/thread:115
    #21 0x7fc29a97de5d  (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xd0e5d)
    #22 0x7fc29ac976b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)
~~~~

**Comment by mtrberzi** [comment]
I'm unfortunately not reproducing this error in ASan in either the debug or release configuration. Could you show me the command-line arguments you're passing to CMake and when you run Z3? It would likely help me reproduce this issue more easily. Thanks!

**Comment by rainoftime** [comment]
Seems fixed in commit 73fa5995d
